File: src/main/javabak/network/IOChangePacket.java
package redcrafter07.processed.network;

import io.netty.buffer.ByteBuf;
import net.minecraft.core.BlockPos;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.neoforged.neoforge.network.handling.IPayloadContext;
import redcrafter07.processed.ProcessedMod;
import redcrafter07.processed.block.machine_abstractions.BlockSide;
import redcrafter07.processed.block.machine_abstractions.IoState;
import redcrafter07.processed.block.machine_abstractions.ProcessedMachine;

import java.io.IOException;


public record IOChangePacket(BlockPos block, IoState state, BlockSide side, boolean itemOrFluid) implements
        CustomPacketPayload {
    public static final CustomPacketPayload.Type<IOChangePacket> TYPE = new CustomPacketPayload.Type<>(ProcessedMod.rl("p_io_change"));
    public static final StreamCodec<ByteBuf, IOChangePacket> CODEC = StreamCodec.composite(
            BlockPos.STREAM_CODEC,
            IOChangePacket::block,
            IoState.STREAM_CODEC,
            IOChangePacket::state,
            BlockSide.STREAM_CODEC,
            IOChangePacket::side,
            ByteBufCodecs.BOOL,
            IOChangePacket::itemOrFluid,
            IOChangePacket::new
    );

    public void handleServer(IPayloadContext context) {
        try (var level = context.player().level()) {
            if (level.getBlockEntity(block) instanceof ProcessedMachine blockEntity) {
                blockEntity.setSide(itemOrFluid, side, state);
                blockEntity.invalidateCapabilities();
            }
        } catch (IOException ignored) {}
    }

    @Override
    public CustomPacketPayload.Type<IOChangePacket> type() {
        return TYPE;
    }
}

File: src/main/kotlin/redcrafter07/processed/network/IOChangePacket.kt
package redcrafter07.processed.network

import io.netty.buffer.ByteBuf
import net.minecraft.core.BlockPos
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.neoforged.neoforge.network.handling.IPayloadContext
import redcrafter07.processed.block.machine_abstractions.BlockSide
import redcrafter07.processed.block.machine_abstractions.IoState
import redcrafter07.processed.block.machine_abstractions.ProcessedMachine
import redcrafter07.processed.rl
import java.io.IOException

data class IOChangePacket(val block: BlockPos, val state: IoState, val side: BlockSide, val itemOrFluid: Boolean) :
    CustomPacketPayload {
    companion object {
        val TYPE: CustomPacketPayload.Type<IOChangePacket> = CustomPacketPayload.Type(rl("p_io_change"))
        val CODEC: StreamCodec<ByteBuf, IOChangePacket> = StreamCodec.composite(
            BlockPos.STREAM_CODEC,
            IOChangePacket::block,
            IoState.STREAM_CODEC,
            IOChangePacket::state,
            BlockSide.STREAM_CODEC,
            IOChangePacket::side,
            ByteBufCodecs.BOOL,
            IOChangePacket::itemOrFluid,
            ::IOChangePacket
        )
    }

    override fun type(): CustomPacketPayload.Type<IOChangePacket> = TYPE

    fun handleServer(context: IPayloadContext) {
        try {
            val level = context.player().level()
            val blockEntity = level.getBlockEntity(block)
            if (blockEntity is ProcessedMachine) {
                blockEntity.setSide(itemOrFluid, side, state)
                blockEntity.invalidateCapabilities()
            }
        } catch (_: IOException) {
        }
    }
}
